File: frontend/src/lib/api/modules.ts
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL ||
  'http://localhost:8000/api/v1';

  export interface Module {
    module_id: number;
    user_id: string; // UUID
    progress: number;
  }

  export interface CreateModuleRequest {
    user_id: string; // UUID
    progress: number;
  }

  export interface UpdateModuleProgressRequest {
    progress: number;
  }

  export interface UserModulesResponse {
    user_id: string; // UUID
    modules: Module[];
  }

  // Get all modules for a specific user
  export async function getUserModules(userId: string): 
  Promise<UserModulesResponse> {
    const response = await fetch(`${API_BASE_URL}/modules/user/${userId}`);
    if (!response.ok) {
      throw new Error('Failed to fetch user modules');
    }
    return response.json();
  }

  // Get a specific module by ID
  export async function getModule(moduleId: number): Promise<Module> {
    const response = await fetch(`${API_BASE_URL}/modules/${moduleId}`);
    if (!response.ok) {
      throw new Error('Failed to fetch module');
    }
    return response.json();
  }

  // Create a new module for a user
  export async function createModule(request: CreateModuleRequest): 
  Promise<Module> {
    const response = await fetch(`${API_BASE_URL}/modules`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to create module');
    }

    return response.json();
  }

  // Update module progress
  export async function updateModuleProgress(
    moduleId: number,
    progress: number
  ): Promise<Module> {
    const response = await fetch(`${API_BASE_URL}/modules/${moduleId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ progress: progress }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to update module progress');
    }

    return response.json();
  }

  // Delete a module
  export async function deleteModule(moduleId: number): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/modules/${moduleId}`, {
      method: 'DELETE',
    });

    if (!response.ok) {
      throw new Error('Failed to delete module');
    }
  }

File: backend/src/app/api/v1/modules.py
from fastapi import APIRouter, HTTPException, Depends
from typing import List
from supabase import Client

from app.core.database import get_supabase
from app.domain.schemas import Module, CreateModuleRequest, UpdateModuleProgressRequest, UserModulesResponse

router = APIRouter()


@router.get("/user/{user_id}", response_model=UserModulesResponse)
async def get_user_modules(user_id: int, db: Client = Depends(get_supabase)):
    """Get all modules for a user"""
    response = db.table("modules").select("*").eq("user_id", user_id).execute()

    modules = [Module(**module) for module in response.data]

    return UserModulesResponse(user_id=user_id, modules=modules)


@router.get("/{module_id}", response_model=Module)
async def get_module(module_id: int, db: Client = Depends(get_supabase)):
    """Get a specific module by ID"""
    response = db.table("modules").select("*").eq("module_id", module_id).execute()

    if not response.data:
        raise HTTPException(status_code=404, detail="Module not found")

    return Module(**response.data[0])


@router.post("", response_model=Module, status_code=201)
async def create_module(request: CreateModuleRequest, db: Client = Depends(get_supabase)):
    """Create a new module for a user"""
    # Check if user exists
    user_response = db.table("users").select("user_id").eq("user_id", request.user_id).execute()

    if not user_response.data:
        raise HTTPException(status_code=404, detail="User not found")

    # Create module
    response = db.table("modules").insert({
        "user_id": request.user_id,
        "module_progress": request.module_progress
    }).execute()

    if not response.data:
        raise HTTPException(status_code=400, detail="Failed to create module")

    return Module(**response.data[0])


@router.put("/{module_id}", response_model=Module)
async def update_module_progress(
    module_id: int,
    request: UpdateModuleProgressRequest,
    db: Client = Depends(get_supabase)
):
    """Update module progress"""
    # Check if module exists
    module_response = db.table("modules").select("*").eq("module_id", module_id).execute()

    if not module_response.data:
        raise HTTPException(status_code=404, detail="Module not found")

    # Validate progress is between 0 and 100
    if request.module_progress < 0 or request.module_progress > 100:
        raise HTTPException(status_code=400, detail="Progress must be between 0 and 100")

    # Update progress
    response = db.table("modules").update({
        "module_progress": request.module_progress
    }).eq("module_id", module_id).execute()

    if not response.data:
        raise HTTPException(status_code=400, detail="Failed to update module")

    return Module(**response.data[0])


@router.delete("/{module_id}", status_code=204)
async def delete_module(module_id: int, db: Client = Depends(get_supabase)):
    """Delete a module"""
    # Check if module exists
    module_response = db.table("modules").select("*").eq("module_id", module_id).execute()

    if not module_response.data:
        raise HTTPException(status_code=404, detail="Module not found")

    # Delete module
    db.table("modules").delete().eq("module_id", module_id).execute()

    return None
